src/Algebra/Module/Morphism/Linear/Properties.agda

```diff
@@ -0,0 +1,127 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Properties of linear maps.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra                   using (CommutativeRing)
+open import Algebra.Module            using (Module)
+open import Level                     using (Level)
+
+module Algebra.Module.Morphism.Linear.Properties
+  {r ℓr m ℓm : Level}
+  {ring      : CommutativeRing r ℓr}
+  (modA modB  : Module ring m ℓm)
+  where
+
+import Algebra.Module.Properties            as ModuleProperties
+import Algebra.Module.Morphism.Structures   as MorphismStructures
+
+open import Axiom.DoubleNegationElimination
+open import Data.Product
+open import Relation.Binary.Reasoning.MultiSetoid
+open import Relation.Nullary          using (¬_)
+open import Relation.Nullary.Negation using (contraposition)
+
+open MorphismStructures.ModuleMorphisms modA modB
+module A = Module modA
+open A using () renaming (Carrierᴹ to A)
+module B = Module modB
+open B using () renaming (Carrierᴹ to B)
+module PA = ModuleProperties modA
+module PB = ModuleProperties modB
+open CommutativeRing ring renaming (Carrier to S)
+
+module _
+  {f : A → B}
+  (isModuleHomomorphism : IsModuleHomomorphism f)
+  where
+
+  open IsModuleHomomorphism isModuleHomomorphism
+
+  x≈0⇒fx≈0 : ∀ {x} → x A.≈ᴹ A.0ᴹ → f x B.≈ᴹ B.0ᴹ
+  x≈0⇒fx≈0 {x} x≈0 = begin⟨ B.≈ᴹ-setoid ⟩
+    f x    ≈⟨ ⟦⟧-cong x≈0 ⟩
+    f A.0ᴹ ≈⟨ 0ᴹ-homo ⟩
+    B.0ᴹ   ∎
+
+  fx≉0⇒x≉0 : ∀ {x} → f x B.≉ᴹ B.0ᴹ → x A.≉ᴹ A.0ᴹ
+  fx≉0⇒x≉0 = contraposition x≈0⇒fx≈0
+
+  -- Zero is a unique output of non-trivial (i.e. - ≉ `const 0`) linear map.
+  zero-unique : ∀ {x} →
+                Σ[ (s , y) ∈ S × A ] ((s A.*ₗ x A.≈ᴹ y) × (f y B.≉ᴹ B.0ᴹ)) →
+                x A.≉ᴹ A.0ᴹ → f x B.≉ᴹ B.0ᴹ
+  zero-unique {x} ((s , y) , (s·x≈y , fy≉0)) x≉0 =
+    PB.s*v≉0⇒v≉0 s·fx≉0
+    where
+    y≉0     : y A.≉ᴹ A.0ᴹ
+    y≉0     = fx≉0⇒x≉0 fy≉0
+    s·fx≉0  : (s B.*ₗ f x) B.≉ᴹ B.0ᴹ
+    s·fx≉0 s·fx≈0 =
+      fy≉0 ( begin⟨ B.≈ᴹ-setoid ⟩
+             f y          ≈⟨ ⟦⟧-cong (A.≈ᴹ-sym s·x≈y) ⟩
+             f (s A.*ₗ x) ≈⟨ *ₗ-homo s x ⟩
+             s B.*ₗ f x   ≈⟨ s·fx≈0 ⟩
+             B.0ᴹ         ∎
+           )
+
+  -- f is odd (i.e. - f (-x) ≈ - (f x)).
+  fx+f[-x]≈0 : ∀ {x} → f x B.+ᴹ f (A.-ᴹ x) B.≈ᴹ B.0ᴹ
+  fx+f[-x]≈0 {x} = begin⟨ B.≈ᴹ-setoid ⟩
+    f x B.+ᴹ f (A.-ᴹ x) ≈⟨ B.≈ᴹ-sym (+ᴹ-homo x (A.-ᴹ x)) ⟩
+    f (x A.+ᴹ (A.-ᴹ x)) ≈⟨ ⟦⟧-cong (A.-ᴹ‿inverseʳ x) ⟩
+    f A.0ᴹ              ≈⟨ 0ᴹ-homo ⟩
+    B.0ᴹ                ∎
+
+  f[-x]≈-fx : ∀ {x} → f (A.-ᴹ x) B.≈ᴹ B.-ᴹ f x
```

Comment:
I suspect the argument `x` should be explicit rather than implicit here, as there's no subsequent argument that Agda can reliably infer it from. The current setup will work sometimes, but not always.

Reply:
Ah, got it; thanks!

**==> Code changed, as requested, in next push.**

**Question:** Should `x` and `y` in the following two functions be made explicit, as well?

Reply:
I think they should be fine as they're provided by an argument :+1:  Our definition of `Injective` in the library has the same form and we don't have any inference problems with that.